includes/admin/upgrades/downgrades.php

```diff
@@ -0,0 +1,151 @@
+<?php
+/**
+ * Downgrades
+ *
+ * @package   easy-digital-downloads
+ * @copyright Copyright (c) 2021, Sandhills Development, LLC
+ * @license   GPL2+
+ * @since     2.11
+ */
+
+/**
+ * Checks if the current site has downgraded, and if so, sets any necessary flags
+ * and updates the version number.
+ *
+ * @since 2.11
+ * @return bool Whether or not a downgrade was performed.
+ */
+function edd_do_downgrade() {
+	$did_downgrade   = false;
+	$edd_version     = preg_replace( '/[^0-9.].*/', '', get_option( 'edd_version' ) );
+	$downgraded_from = get_option( 'edd_version_downgraded_from' );
+
+	/**
+	 * Check for downgrade from 3.0 to 2.11.
+	 */
+	if ( version_compare( EDD_VERSION, '3.0-beta1', '<' ) ) {
+		if (
+			version_compare( $edd_version, '3.0-beta1', '>=' ) ||
+			( $downgraded_from && version_compare( $downgraded_from, '3.0-beta1', '>=' ) )
+		) {
+			/*
+			 * This site probably just downgraded from EDD 3.0. Let's store a flag so we can give them the option to properly roll back.
+			 * @see edd_show_downgrade_notices()
+			 */
+			update_option( 'edd_v3_downgrade', time() );
+			$did_downgrade = true;
+		}
+	}
+
+	if ( $did_downgrade ) {
+		update_option( 'edd_version', preg_replace( '/[^0-9.].*/', '', EDD_VERSION ) );
+		delete_option( 'edd_version_downgraded_from' );
+	}
+
+	return $did_downgrade;
+}
+add_action( 'admin_init', 'edd_do_downgrade' );
+
+/**
+ * Display downgrade notices.
+ *
+ * @since 2.11
+ */
+function edd_show_downgrade_notices() {
+	if ( ! current_user_can( 'manage_options' ) ) {
+		return;
+	}
+
+	/**
+	 * EDD 3.0 downgrade.
+	 */
+	if ( get_option( 'edd_v3_downgrade' ) ) {
+		$downgrade_url = wp_nonce_url( add_query_arg( 'edd_action', 'downgrade_v3' ), 'edd_downgrade_v3' );
+		$dismiss_url   = wp_nonce_url( add_query_arg( 'edd_action', 'dismiss_v3_downgrade' ), 'edd_dismiss_v3_downgrade' );
+		?>
+		<div class="notice notice-warning">
+			<h2><?php esc_html_e( 'Did you downgrade from Easy Digital Downloads 3.0?', 'easy-digital-downloads' ); ?></h2>
+			<p>
+				<?php _e( 'We\'ve detected that your site may have just downgraded from Easy Digital Downloads version 3.0. If that is correct, please click the "Complete Downgrade" button below to complete this process.', 'easy-digital-downloads' ); ?>
```

Comment:
```suggestion
				<?php esc_html_e( 'We\'ve detected that your site may have just downgraded from Easy Digital Downloads version 3.0. If that is correct, please click the "Complete Downgrade" button below to complete this process. This will change your database, but is necessary to use this version of Easy Digital Downloads.', 'easy-digital-downloads' ); ?>
```
Maybe? That way they know at least a little bit amount what it's doing, and indicates that they can't use their store without doing it?

Reply:
I didn't want people to be afraid to click it. :sweat_smile: I'm in two minds about it. I want to treat it like no big deal, because it really is no big deal. Does `change your database` sound scary?

Under other circumstances this is the kind of update we'd do behind the scenes without even telling the user about it. (Which is what we do on upgrade to 3.0. We make this change before you even run the migration.)

Reply:
Yes, "change your database" does sound a bit scary, and I understand it's not a destructive change. I can live without it, but I do think the message needs to communicate that their store will not operate correctly in 2.x if they don't click the button.

> Under other circumstances this is the kind of update we'd do behind the scenes without even telling the user about it.

Is that something to consider here instead maybe?

Reply:
> Is that something to consider here instead maybe?

I think originally we didn't because we were kind of guessing if it was a downgrade, and it was less certain than an upgrade. But now that it's done... I think we can be pretty certain. I'll think on it for an hour. :stuck_out_tongue:

Reply:
If we are leaving the message in so they have to trigger the downgrade manually, should the notice say something like, "You must complete the downgrade to successfully process orders in version 2.11." ?

I guess my concern is just that someone might downgrade and then ignore the notice. Functionally, the PR is working for me and I can approve if we don't want to tweak this part.

Reply:
I'll rework it to happen automatically.